spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-spi/src/main/java/org/springframework/cloud/data/module/ModuleStatus.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2014 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.module;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.springframework.xd.module.ModuleDescriptor;
+
+/**
+ * Status of a {@link ModuleDescriptor} deployment. This status is
+ * composed of an aggregate of all individual module deployments.
+ * <p>
+ * Consumers of the SPI obtain module status via
+ * {@link org.springframework.cloud.data.module.deployer.ModuleDeployer#status},
+ * whereas SPI implementations create instances of this class via
+ * {@link ModuleStatus.Builder}.
+ *
+ * @see ModuleInstanceStatus
+ *
+ * @author Patrick Peralta
+ */
+public class ModuleStatus {
+
+	/**
+	 * Module deployment states. Unless indicated, these states
+	 * may represent the state of an individual module deployment
+	 * or the aggregate state of all module instances for a given
+	 * {@link ModuleDescriptor}.
+	 */
+	public enum State {
+
+		/**
+		 * The module is being deployed. If there are multiple modules,
+		 * at least one module is being deployed.
+		 */
+		deploying,
+
+		/**
+		 * All modules have been deployed.
+		 */
+		deployed,
+
+		/**
+		 * In the case of multiple modules, some have successfully deployed.
+		 * This state does not apply for individual modules.
+		 */
+		incomplete,
```

Comment:
Given this is per module status, I am trying to understand why do we need `incomplete` status here and under what scenario we would update the ModuleStatus to `incomplete` at the **ModuleDeployer** level. Based on what I see, the deployment happens per module descriptor.